[RISCV] Keep same SEW/LMUL ratio if possible in forward transfer

lukel97 opened the pull request:
For instructions like vmv.s.x and friends where we don't care about LMUL or the
SEW/LMUL ratio, we can change the LMUL in its state so that it has the same
SEW/LMUL ratio as the previous state. This allows us to avoid more VL toggles
later down the line (i.e. use vsetvli zero, zero, which requires that the
SEW/LMUL ratio must be the same)

This is an alternative approach to the idea in #69259, but note that they 
don't catch exactly the same test cases.

lukel97 committed:
[RISCV] Keep same SEW/LMUL ratio if possible in forward transfer

For instructions like vmv.s.x and friends where we don't care about LMUL or the
SEW/LMUL ratio, we can change the LMUL in its state so that it has the same
SEW/LMUL ratio as the previous state. This allows us to avoid more VL toggles
later down the line (i.e. use vsetvli zero, zero, which requires that the
SEW/LMUL ratio must be the same)

lukel97 commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-reduction-fp.ll:
```diff
@@ -611,11 +607,10 @@ define float @vreduce_fwadd_v32f32(ptr %x, float %s) {
 ; CHECK-NEXT:    li a1, 32
-; CHECK-NEXT:    vsetvli zero, a1, e16, m4, ta, ma
+; CHECK-NEXT:    vsetvli zero, a1, e32, m8, ta, ma
 ; CHECK-NEXT:    vle16.v v8, (a0)
```
Interesting side effect I didn't expect: This changes the VTYPE of unit-stride loads since they don't use LMUL, just the SEW/LMUL ratio.

preames commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int.ll:
```diff
@@ -1101,21 +1101,20 @@ define void @urem_v2i64(ptr %x, ptr %y) {
 ; CHECK:       # %bb.0:
-; CHECK-NEXT:    vsetivli zero, 16, e8, m1, ta, ma
+; CHECK-NEXT:    vsetivli zero, 16, e16, m2, ta, ma
 ; CHECK-NEXT:    vle8.v v8, (a0)
```
Actually, I'm wrong here.  Ignore this.

The alignment is not tied to LMUL for these instructions, it's tied to EMUL.  The EMUL of the load hasn't changed here, and thus this is still fine.

lukel97 replied:
Yeah this test changed surprised me, it seems to be taking advantage of this existing bit of logic in the backward pass:
https://github.com/llvm/llvm-project/blob/46028022407d003b8af4ddf27a4679de4891f10d/llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp#L372-L381

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1042,6 +1042,26 @@ void RISCVInsertVSETVLI::transferBefore(VSETVLIInfo &Info,
+      bool Fractional = NewRatio < 8;
+      RISCVII::VLMUL NewVLMul = RISCVVType::encodeLMUL(
+          Fractional ? 8 / NewRatio : NewRatio / 8, Fractional);
+      unsigned VType = Info.encodeVTYPE();
```
Given SEW is unchanged here, I think just adding a setVLMul accessor is probably cleaner.

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1042,6 +1042,26 @@ void RISCVInsertVSETVLI::transferBefore(VSETVLIInfo &Info,
+  // maintain the SEW/LMUL ratio. This allows us to eliminate VL toggles in more
+  // places.
+  DemandedFields Demanded = getDemanded(MI, MRI);
+  if (!Demanded.LMUL && !Demanded.SEWLMULRatio && Info.isValid() &&
```
I mentioned this offline, but basing this off the existing demanded feilds is much better than re-infering which is what I'd tried   This makes cornercase bugs around instructions (e.g. vrgather, or vslide*) much less likely because the existing getDemandedFeidls logic would have to be wrong as well.  I feel a bit silly in retrospect, but glad you got it structured properly.  :)

preames commented on llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp:
```diff
@@ -1042,6 +1042,26 @@ void RISCVInsertVSETVLI::transferBefore(VSETVLIInfo &Info,
+      PrevInfo.isValid() && !Info.isUnknown() && !PrevInfo.isUnknown() &&
+      !Info.hasSameVLMAX(PrevInfo)) {
+    unsigned SEW = Info.getSEW();
+    // Fixed point value with 3 fractional bits.
```
This looks very similar to getSameRatioLMUL, but with an extra bounds check.  Could we manage to share code somehow?  Maybe return an optional and handle the error case in MCA?  Since I think, at least in principle, the error case is possible there?

preames reviewed (commented):
I think this is correct, a couple comments to cleanup the code for submission.

lukel97 committed:
Update test outside of rvv directory

lukel97 committed:
Refactor getSameRatioLMUL so we can reuse it, add setVLMul accessor

lukel97 committed:
Remove same VLMAX check

If they have the same VLMAX then adjusting the SEW/LMUL ratio is idempotent
anyway. Removing the check to simplify it a bit.

preames commented on llvm/test/CodeGen/RISCV/rvv/vreductions-fp-vp.ll:
```diff
@@ -342,14 +342,14 @@ define half @vpreduce_ord_fadd_nxv64f16(half %s, <vscale x 64 x half> %v, <vscal
 ; ZVFHMIN-NEXT:    fcvt.h.s fa5, fa5
 ; ZVFHMIN-NEXT:    fcvt.s.h fa5, fa5
-; ZVFHMIN-NEXT:    vsetivli zero, 1, e32, m1, ta, ma
+; ZVFHMIN-NEXT:    vsetivli zero, 1, e32, m8, ta, ma
```
Just to flag this - this points out that we're changing LMUL without considering whether having the same ratio is enough to actually emit the vl preserving form.  I'm noting this for the record, but I do not think this is a practical problem.

preames commented on llvm/test/CodeGen/RISCV/rvv/fixed-vectors-masked-gather.ll:
```diff
@@ -1259,15 +1260,16 @@ define <4 x i16> @mgather_truemask_v4i16(<4 x ptr> %ptrs, <4 x i16> %passthru) {
-; RV64ZVE32F-NEXT:    vsetivli zero, 2, e16, mf2, tu, ma
+; RV64ZVE32F-NEXT:    vsetvli zero, zero, e16, m1, ta, ma
 ; RV64ZVE32F-NEXT:    vmv.s.x v9, a2
+; RV64ZVE32F-NEXT:    vsetivli zero, 2, e16, mf2, tu, ma
```
To flag - we're actually adding a vsetvli which didn't exist.  This is arguably a regression, but a pretty minor one since the "added" vsetvli is effectively a vtype toggle (i.e. the line I commented on "moved").  I think this also goes away once the patch on review for doing this in the backwards pass lands.

preames reviewed (approved):
LGTM

wangpc-pp commented:
Just curious about why `llvm/unittests/Target/RISCV/RISCVBaseInfoTest.cpp` is passed? Is `gtest` able to auto dereference `std::optional`?

lukel97 commented:
> Just curious about why `llvm/unittests/Target/RISCV/RISCVBaseInfoTest.cpp` is passed? Is `gtest` able to auto dereference `std::optional`?

std::optional has equality defined on it so that `make_optional(x) == x`, I presume that's why the tests didn't need updated
